src/storage/secondary/RedisStorage.cpp

```diff
@@ -0,0 +1,374 @@
+parse_connect_timeout(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("connect-timeout");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return milliseconds_to_timeval(it->second);
+}
+
+static nonstd::optional<struct timeval>
+parse_operation_timeout(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("operation-timeout");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return milliseconds_to_timeval(it->second);
+}
+
+static nonstd::optional<std::string>
+parse_username(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("username");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+static nonstd::optional<std::string>
+parse_password(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("password");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+RedisStorage::RedisStorage(const Url& url, const AttributeMap& attributes)
+  : m_url(url),
+    m_connect_timeout(parse_connect_timeout(attributes)),
+    m_operation_timeout(parse_operation_timeout(attributes)),
+    m_username(parse_username(attributes)),
+    m_password(parse_password(attributes))
+{
+  m_prefix = "ccache"; // TODO: attribute
+  m_context = nullptr;
+  m_connected = false;
+  m_invalid = false;
+}
+
+RedisStorage::~RedisStorage()
+{
+  disconnect();
+  if (m_context) {
+    redisFree(m_context);
+    m_context = nullptr;
+  }
+}
+
+int
+RedisStorage::connect()
+{
+  if (m_connected) {
+    return REDIS_OK;
+  }
+  if (m_invalid) {
+    return REDIS_ERR;
+  }
+
+  if (m_context) {
+    if (redisReconnect(m_context) == REDIS_OK) {
+      m_connected = true;
+      return REDIS_OK;
+    }
+    LOG("Redis reconnect err: {}", m_context->errstr);
+    redisFree(m_context);
+    m_context = nullptr;
+  }
+
+  ASSERT(m_url.scheme() == "redis");
+  std::string host = m_url.host();
+  std::string port = m_url.port();
+  std::string sock = m_url.path();
+  if (m_connect_timeout) {
+    LOG("Redis connect timeout {}", timeval_to_string(*m_connect_timeout));
+  }
+  if (!host.empty()) {
+    int p = port.empty() ? 6379 : std::stoi(port);
+    if (m_connect_timeout) {
+      m_context = redisConnectWithTimeout(host.c_str(), p, *m_connect_timeout);
+    } else {
+      m_context = redisConnect(host.c_str(), p);
+    }
+  } else if (!sock.empty()) {
+    if (m_connect_timeout) {
+      m_context = redisConnectUnixWithTimeout(sock.c_str(), *m_connect_timeout);
+    } else {
+      m_context = redisConnectUnix(sock.c_str());
+    }
+  } else {
+    LOG("Redis invalid url: {}", m_url.str());
+    m_invalid = true;
+    return REDIS_ERR;
+  }
+
+  if (!m_context) {
+    LOG("Redis connect {} err NULL", m_url.str());
+    m_invalid = true;
+    return REDIS_ERR;
+  } else if (m_context->err) {
+    LOG("Redis connect {} err: {}", m_url.str(), m_context->errstr);
```

Comment:
I tried building with a locally compiled libhiredis 1.0.0 (and also with latest hiredis master) and I always get an error with `m_context->err` and `m_context->errstr` being different garbage values. Have you tried hiredis 1.0.0? I can debug this when I get time, maybe tomorrow.

Reply:
This is most likely false alarm – I seem to somehow have used a mix-up of header and library files of different versions.

Reply:
I have tried with 0.13.3 (18.04), 0.14.0 (20.04) and 1.0.0 (brew). But maybe some more asserts etc, are needed ?